sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactoryExpressionJsonConverter.cs

```diff
@@ -2,41 +2,210 @@
+                genericType = type;
+                return true;
+            }
+
+            genericType = null;
+            return false;
+        }
+
+        private static bool IsGenericListType(Type type) => type == typeof(IList<>);
+
+        private static void Serialize<T>(Utf8JsonWriter writer, DataFactoryExpression<T?> expression)
+        {
+            if (expression.HasLiteral)
+            {
+                writer.WriteObjectValue(expression.LiteralInternal!);
+            }
+            else
+            {
+                SerializeExpression(writer, expression.Type!, expression.Expression!);
+            }
+        }
+
+        private static void SerializeList<T>(Utf8JsonWriter writer, DataFactoryExpression<IList<T?>> expression)
+        {
+            if (expression.HasLiteral && expression.Literal == null)
+            {
+                writer.WriteNullValue();
+                return;
+            }
+            if (expression.HasLiteral)
+            {
+                writer.WriteStartArray();
+                foreach (T? elem in expression.Literal!.AsEnumerable())
+                {
+                    JsonSerializer.Serialize(writer, elem);
+                }
+                writer.WriteEndArray();
+            }
+            else
+            {
+                SerializeExpression(writer, expression.Type!, expression.Expression!);
+            }
+        }
+
+        private static void SerializeExpression(Utf8JsonWriter writer, string type, string value)
         {
-            writer.WriteObjectValue(value);
+            writer.WriteStartObject();
+            writer.WritePropertyName("type");
+            writer.WriteStringValue(type);
+            writer.WritePropertyName("value");
+            writer.WriteStringValue(value);
+            writer.WriteEndObject();
+        }
+
+        private static DataFactoryExpression<IList<T?>> DeserializeList<T>(JsonElement element)
+        {
+            if (element.ValueKind == JsonValueKind.Array)
+            {
+                var list = new ChangeTrackingList<T?>();
+                foreach (var item in element.EnumerateArray())
+                {
+                    list.Add(item.ValueKind == JsonValueKind.Null ? default : JsonSerializer.Deserialize<T>(item.ToString()!));
+                }
+
+                return new DataFactoryExpression<IList<T?>>(list);
+            }
+
+            if (element.ValueKind == JsonValueKind.Object)
+            {
+                return DeserializeExpression<IList<T?>>(element);
+            }
+
+            throw new InvalidOperationException($"Cannot deserialize an {element.ValueKind} as a list.");
+        }
+
+        internal static DataFactoryExpression<T>? Deserialize<T>(JsonElement element)
+        {
+            T? value = default;
+
+            if (element.ValueKind == JsonValueKind.Null)
+            {
+                return null;
+            }
+
+            if (element.ValueKind == JsonValueKind.Object)
```

Comment:
For now at least, we have the invariant that Json Object => expression. In theory this could change so we would probably need to inspect the payload to determine if it is an expression or not.

Reply:
Lets leave a TODO here please